### PR TITLE
change(common/models): model-template package conversion to ES modules 🧩

### PR DESCRIPTION
**Pull request by jahorton**
This PR converts our predictive-text lexical-model template module to use of ES modules.

So, the part that interprets our compiled Trie-based dictionaries for actual use as functioning lexical models.

Also, adds a `build.sh` script using our relatively new "builder" format, allowing integration of it into dependency checks found in higher-level packages.

### CODE REVIEW

File: common/models/templates/build.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+## START STANDARD BUILD SCRIPT INCLUDE
+# adjust relative paths as necessary
+THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BASH_SOURCE[0]}")"
+. "$(dirname "$THIS_SCRIPT")/../../../resources/build/build-utils.sh"
+. "$KEYMAN_ROOT/resources/shellHelperFunctions.sh"
+## END STANDARD BUILD SCRIPT INCLUDE
+
+# This script runs from its own folder
+cd "$(dirname "$THIS_SCRIPT")"
```
Comment (mcdurdin) [inline common/models/templates/build.sh:15]: This is going to start to look familiar ... :grin:

```suggestion
## END STANDARD BUILD SCRIPT INCLUDE

. "$KEYMAN_ROOT/resources/shellHelperFunctions.sh"

# This script runs from its own folder
cd "$THIS_SCRIPT_PATH"
```

File: common/models/templates/build.sh [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+################################ Main script ################################
+
+# TODO: for predictive-text, we only need :headless, perhaps we should be splitting modules?
+# TODO: remove :tools once kmlmc is a dependency for test:module
+
```
Comment (mcdurdin) [inline common/models/templates/build.sh:21]:
```suggestion
```

File: common/models/templates/src/common.ts [hunk 3]
```diff
@@ -1,148 +1,153 @@
...
-
-    return {
-      left: leftContext,
-      right: rightContext,
-      startOfBuffer: context.startOfBuffer,
-      endOfBuffer: context.endOfBuffer,
-      casingForm: context.casingForm
-    };
-  }
+/// <reference types="@keymanapp/models-types" />
```
Comment (mcdurdin) [inline common/models/templates/src/common.ts:1]: I thought modules allowed us to eliminate `/// <reference`s?

The [docs](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html#-reference-types-) say:

> Use these directives only when you’re authoring a `d.ts` file by hand.

File: common/models/templates/src/common.ts [hunk 4]
```diff
@@ -1,148 +1,153 @@
...
-   */
-  export function isLowSurrogate(codeUnit: number): boolean;
-  export function isLowSurrogate(codeUnit: string|number): boolean {
-    if(typeof codeUnit == 'string') {
-      codeUnit = codeUnit.charCodeAt(0);
+  let leftContext = fullLeftContext.kmwSubstr(0, lLen - lDel) + (transform.insert || '');
+
+  let fullRightContext = context.right || '';
+  let rLen = fullRightContext.kmwLength();
+  let rDel = (rLen < (transform.deleteRight ?? 0)) ? rLen : (transform.deleteRight ?? 0);
```
Comment (mcdurdin) [inline common/models/templates/src/common.ts:20]: Is this a separate bug fix?

File: common/models/templates/src/priority-queue.ts [hunk 5]
```diff
@@ -4,216 +4,215 @@
...
-        let index = queuedIndices.shift();
-        let parent = PriorityQueue.parentIndex(index);
-        if(this.siftDown(index) && parent >= 0 && lastParent != parent) {
-            // We only need to queue examination for a heap node if its children have changed.
-          queuedIndices.push(parent);
-          lastParent = parent;
-        }
+    lastParent = -1;
+    while(queuedIndices.length > 0) {
+      let index = queuedIndices.shift() as number;
```
Comment (mcdurdin) [inline common/models/templates/src/priority-queue.ts:83]: Why is this change required?

File: common/models/templates/src/priority-queue.ts [hunk 6]
```diff
@@ -4,216 +4,215 @@
...
+    this.heapify(firstParent >= 0 ? firstParent : 0, PriorityQueue.parentIndex(this.count-1));
+  }
 
-      return root;
+  /**
+   * Removes the highest-priority element from the queue, returning it.
+   * <p>
+   * Is O(log N), where N = number of items in the priority queue.
+   */
+  dequeue(): Type | undefined {
```
Comment (mcdurdin) [inline common/models/templates/src/priority-queue.ts:157]: Why is this change required?

File: common/models/templates/src/priority-queue.ts [hunk 7]
```diff
@@ -4,216 +4,215 @@
...
-    private siftDown(index: number): boolean {
-      let leftIndex = PriorityQueue.leftChildIndex(index);
-      let rightIndex = PriorityQueue.rightChildIndex(index);
-      let topMostIndex = index;
-
-      if(leftIndex < this.heap.length && this.comparator(this.heap[leftIndex], this.heap[topMostIndex]) < 0) {
-        topMostIndex = leftIndex;
-      }
+    const root = this.heap[0];
+    let tail = this.heap.pop() as Type;
```
Comment (mcdurdin) [inline common/models/templates/src/priority-queue.ts:163]: Why is this change required?

File: common/models/templates/src/priority-queue.ts [hunk 8]
```diff
@@ -4,216 +4,215 @@
...
+   * - O(N) when 'heapifying' the whole heap
+   * - O(N) worst-case for partial heap operations (as part of an enqueueAll)
+   * <p>
+   */
+  private heapify(): void;
+  private heapify(start: number, end: number): void;
+  private heapify(start?: number, end?: number): void {
+    if(start == undefined || end == undefined) {
+      this.heapify(0, this.count - 1);
+      return;
```
Comment (mcdurdin) [inline common/models/templates/src/priority-queue.ts:62]: Why is this change required?

File: common/models/templates/src/tokenization.ts [hunk 9]
```diff
@@ -1,129 +1,129 @@
+/// <reference types="@keymanapp/models-types" />
```
Comment (mcdurdin) [inline common/models/templates/src/tokenization.ts:1]: A bunch of changes in this file don't immediately appear to be related to the change to ESModules?

File: common/models/templates/src/trie-model.ts [hunk 10]
```diff
@@ -24,535 +26,525 @@
...
-     */
-    return wordform
-      .normalize('NFD')
-      // Remove all combining diacritics (if input is in NFD)
-      // common to Latin-orthographies.
-      .replace(/[\u0300-\u036f]/g, '')
-      .toLowerCase() as SearchKey;
-  }
-
-  function getDefaultWordBreaker() {
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:547]: Does eliminating this reduce functionality in any way?

File: common/models/templates/src/trie-model.ts [hunk 11]
```diff
@@ -24,535 +26,525 @@
...
+ * @class TrieModel
+ *
+ * Defines a trie-backed word list model, or the unigram model.
+ * Unigram models throw away all preceding words, and search
+ * for the next word exclusively. As such, they can perform simple
+ * prefix searches within words, however they are not very good
+ * at predicting the next word.
+ */
+export default class TrieModel implements LexicalModel {
+  configuration?: Configuration;
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:96]: I don't understand how this and a number of other changes in this file relate to ESModules?

**Review by mcdurdin — CHANGES_REQUESTED** [review]
A number of changes sprinkled through these files which I don't understand. May be fine but can you explain them for me?

File: common/models/templates/src/tokenization.ts (hunk 9 above)
Comment (jahorton) [reply common/models/templates/src/tokenization.ts:1]: I remember that during some of the modularization work, VS Code's TS engine got a _lot_ stricter about types for a while.  (It particularly didn't like certain things that were happening when `undefined` was possible.)  

I'm not sure if this was prompted by conversion to modules or what, but I figured small changes to satisfy the stricter type checks couldn't hurt.

File: common/models/templates/src/trie-model.ts (hunk 10 above)
Comment (jahorton) [reply common/models/templates/src/trie-model.ts:547]: It's not eliminated - this is handled via `import` now.  (See line 32.)

File: common/models/templates/src/common.ts (hunk 3 above)
Comment (jahorton) [reply common/models/templates/src/common.ts:1]: ... maybe this was also related to the sudden compiler strictness I'd experienced?  I seem to be able to remove it just fine now, at least.

File: common/models/templates/src/priority-queue.ts (hunk 5 above)
Comment (jahorton) [reply common/models/templates/src/priority-queue.ts:83]: image

TypeScript is failing to infer that the `| undefined` part is logically impossible here, which causes the following lines that expect a `number` to throw related errors.

File: common/models/templates/src/priority-queue.ts (hunk 6 above)
Comment (jahorton) [reply common/models/templates/src/priority-queue.ts:157]: Line 159.

> Type 'undefined' is not assignable to type 'Type'.
>  'Type' could be instantiated with an arbitrary type which could be unrelated to 'undefined'.ts(2322)

File: common/models/templates/src/priority-queue.ts (hunk 7 above)
Comment (jahorton) [reply common/models/templates/src/priority-queue.ts:163]: Similar type-strictness shenanigans.  Note lines 158-160:  they logically prevent `.pop()` from returning `undefined`, but the TS engine isn't inferring that detail properly.

File: common/models/templates/src/trie-model.ts (hunk 10 above)
Comment (mcdurdin) [reply common/models/templates/src/trie-model.ts:547]: Yeah, just this function had more than one way to obtain the default?

**Review by mcdurdin — APPROVED** [review]

File: common/models/templates/src/trie-model.ts (hunk 10 above)
Comment (jahorton) [reply common/models/templates/src/trie-model.ts:547]: I _think_ it's because one way found them when compiled into `lm-worker` (where the namespace already exists and we don't want duplication) while the other linked them when testing separately from the `lm-worker` that would establish the namespace?  

(`wordBreakers` was a top-level variable name as a compiled namespace.)

Meanwhile, `import` makes this simpler.
